composer/datasets/streaming/dataset.py

```diff
@@ -0,0 +1,379 @@
+            else:
+                if not self._are_all_shards_downloaded:
+                    self._downloaded_ids.extend(new_ids)
+                for todo_ids in self._epoch_to_todo_ids.values():
+                    todo_ids.extend(new_ids)
+
+    def _load_shards_if_downloaded(self, shards: Sequence[int], part_min_id: int, part_max_id: int) -> List[int]:
+        """Load any of the given shards that are already present in the cache, returning the missing shards.
+
+        Args:
+            shards (Sequence[int]): The shards to attempt to load.
+            part_min_id (int): Minimum sample ID of this partition.
+            part_max_id (int): Maximum sample ID of this partition.
+
+        Returns:
+            list of int: The shards that remain to be loaded.
+        """
+        downloaded = []
+        missing = []
+        for shard in sorted(shards):
+            basename = get_shard_basename(shard)
+            local = os.path.join(self.local, basename)
+            if os.path.exists(local):
+                downloaded.append(shard)
+            else:
+                missing.append(shard)
+        if downloaded:
+            self._load_shards(downloaded, part_min_id, part_max_id)
+        return missing
+
+    def _done_loading(self) -> None:
+        """Callback on completion of loading my shards."""
+        with self._lock:
+            self._are_all_shards_downloaded = True
+
+    def _download_thread(self, shards: Sequence[int], part_min_id: int, part_max_id: int) -> None:
+        """Background thread to download and assimilate missing shards.
+
+        Args:
+            shards (list of int): The shards remaining to be downloaded.
+            part_min_id (int): Minimum sample ID of this partition.
+            part_max_id (int): Maximum sample ID of this partition.
+        """
+        shards = list(shards)
+        if self.shuffle:
+            np.random.shuffle(shards)
+        for shard in shards:
+            basename = get_shard_basename(shard)
+            self._download_if_missing(basename)
+            shards = shard,
+            self._load_shards(shards, part_min_id, part_max_id)
+        self._done_loading()
+
+    def _load(self) -> None:
+        """Load shards."""
+        # We find out num workers, and therefore num partitions, when __iter__ is called.
+        # From the partition, derive our shard overlap range and exact sample range.
+        world = World.from_env()
+        todo_shards, part_min_id, part_max_id = self.index.get_partition(world, self.batch_size)
+
+        # Preload all of our shards that are already available in the cache.
+        todo_shards = self._load_shards_if_downloaded(todo_shards, part_min_id, part_max_id)
+
+        # Start downloading our missing shards in a background thread, if there are any.
+        if todo_shards:
+            thread = Thread(target=self._download_thread, args=(todo_shards, part_min_id, part_max_id), daemon=True)
```

Comment:
Would it make sense to support multiple downloading threads? (If the shard size is small, I presume we would not be bandwidth bottlenecked with just one thread?)

Reply:
Tracking this but not addressing in this PR.

In practice we have many workers per device, and many devices per node, so maybe 32 or 64 download threads per node, which is pretty good about maxing out our bandwidth. Shard size is also defaulted to 16MB, which seems OK, maybe could be bigger? It's a tradeoff of network download efficiency vs. how long you are willing to wait for sample 1.